heron/shell/src/python/utils.py

```diff
@@ -141,34 +143,40 @@ def read_chunk(filename, offset=None, length=None):
 
   return dict(offset=offset, length=0)
 
-def stream_to_string(stream):
-  """
-  Converts stream to string. Blocks until end of stream
-  """
-  str_builder = ''
-  while True:
-    line = stream.readline()
-    if not line:
-      break
-    str_builder += line
-  return str_builder
+class StringBuilder(object):
+  def __init__(self):
+    self.str = ""
+
+  def add(self, line):
+    self.str += line
+
+  def result(self):
+    return self.str
 
-def pipe(in_stream, to_cmd):
+def pipe(prev_proc, to_cmd):
   """
   Pipes in_stream from output of previous pipe into to_cmd.
   Returns stdout stream of to_cmd
   """
-  p = subprocess.Popen(to_cmd,
-                       stdout=subprocess.PIPE,
-                       stdin=in_stream)
-  return p.stdout
+  stdin = None if prev_proc is None else prev_proc.stdout
+  proc = subprocess.Popen(to_cmd,
+                          stdout=subprocess.PIPE,
+                          stdin=stdin)
+  if prev_proc is not None:
+    prev_proc.stdout.close() # Allow prev_proc to receive a SIGPIPE
+  return proc
 
-def str_cmd(cmd):
+def str_cmd(cmd, cwd=None, env=None):
   """
   Runs the command and returns its stdout and stderr.
   """
-  p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-  (stdout, stderr) = (stream_to_string(p.stdout), stream_to_string(p.stderr))
+  proc = subprocess.Popen(cmd, stdout=subprocess.PIPE,
+                          stderr=subprocess.PIPE, cwd=cwd, env=env)
+  stdout_builder, stderr_builder = StringBuilder(), StringBuilder()
+  log.async_stream_process_stdout(proc, stdout_builder.add)
+  log.async_stream_process_stderr(proc, stderr_builder.add)
+  proc.wait()
```

Comment:
If the `proc.wait()` call is blocked forever, then it will never return. And the method is blocked here. Just want to make sure this case won't happen.